Full Performer sync taking place everytime, scenes with excluded tags being shown when sorting, random sort not really random everytime.

Hi, i'm having a few issues.
I'm running the latest 3.0.0 version.
1. On every incremental sync, my performers are being synced fully every time.
```
[2025-12-25T12:11:57.330Z] [INFO] Scheduled incremental sync triggered
[2025-12-25T12:11:57.507Z] [INFO] Starting incremental sync {"since":"2025-12-17T08:18:12.430Z"}
[2025-12-25T12:11:57.507Z] [INFO] Syncing tags...
[2025-12-25T12:11:58.910Z] [INFO] Tags synced: 10 in 1.4s
[2025-12-25T12:11:59.225Z] [INFO] Syncing studios...
[2025-12-25T12:12:00.531Z] [INFO] Studios synced: 18 in 1.3s
[2025-12-25T12:12:00.708Z] [INFO] Syncing performers...
[2025-12-25T12:23:23.594Z] [INFO] Performers synced: 1,947 in 682.7s
[2025-12-25T12:23:24.432Z] [INFO] Syncing groups...
[2025-12-25T12:23:26.862Z] [INFO] Groups synced: 5 in 2.4s
[2025-12-25T12:23:27.003Z] [INFO] Syncing galleries...
[2025-12-25T12:23:27.008Z] [INFO] Galleries synced: 0 in 0.0s
[2025-12-25T12:23:27.222Z] [INFO] Syncing scenes...
[2025-12-25T12:23:39.063Z] [INFO] Scenes synced: 319 in 11.8s
[2025-12-25T12:23:39.253Z] [INFO] Syncing images...
[2025-12-25T12:23:39.286Z] [INFO] Images synced: 0 in 0.0s
[2025-12-25T12:23:39.396Z] [INFO] Rebuilding user stats after sync...
[2025-12-25T12:23:39.396Z] [INFO] Rebuilding stats for all users
[2025-12-25T12:23:39.422Z] [INFO] Rebuilding stats for user {"userId":4}
[2025-12-25T12:23:39.458Z] [INFO] Stats rebuild complete {"userId":4,"performerCount":0,"studioCount":0,"tagCount":0}
[2025-12-25T12:23:39.458Z] [INFO] Rebuilding stats for user {"userId":1}
[2025-12-25T12:24:11.853Z] [INFO] Stats rebuild complete {"userId":1,"performerCount":274,"studioCount":140,"tagCount":694}
[2025-12-25T12:24:11.854Z] [INFO] Rebuilding stats for user {"userId":2}
[2025-12-25T12:24:16.281Z] [INFO] Stats rebuild complete {"userId":2,"performerCount":407,"studioCount":177,"tagCount":886}
[2025-12-25T12:24:16.281Z] [INFO] Rebuilding stats for user {"userId":3}
[2025-12-25T12:24:17.575Z] [INFO] Stats rebuild complete {"userId":3,"performerCount":11,"studioCount":7,"tagCount":121}
[2025-12-25T12:24:17.576Z] [INFO] All stats rebuild complete {"userCount":4}
[2025-12-25T12:24:17.576Z] [INFO] User stats rebuild complete
[2025-12-25T12:24:17.576Z] [INFO] Incremental sync completed {"durationMs":740222,"results":[{"type":"tag","synced":10},{"type":"studio","synced":18},{"type":"performer","synced":1947},{"type":"group","synced":5},{"type":"gallery","synced":0},{"type":"scene","synced":319},{"type":"image","synced":0}]}
[2025-12-25T13:11:57.458Z] [INFO] Scheduled incremental sync triggered
[2025-12-25T13:11:59.736Z] [INFO] Starting incremental sync {"since":"2025-12-17T08:18:12.430Z"}
[2025-12-25T13:11:59.736Z] [INFO] Syncing tags...
[2025-12-25T13:12:05.368Z] [INFO] Tags synced: 10 in 5.2s
[2025-12-25T13:12:06.604Z] [INFO] Syncing studios...
[2025-12-25T13:12:12.320Z] [INFO] Studios synced: 18 in 5.7s
[2025-12-25T13:12:13.342Z] [INFO] Syncing performers...
[2025-12-25T13:28:50.271Z] [INFO] Performers synced: 1,947 in 996.9s
[2025-12-25T13:28:50.578Z] [INFO] Syncing groups...
[2025-12-25T13:28:52.031Z] [INFO] Groups synced: 5 in 1.5s
[2025-12-25T13:28:52.207Z] [INFO] Syncing galleries...
[2025-12-25T13:28:52.216Z] [INFO] Galleries synced: 0 in 0.0s
[2025-12-25T13:28:52.359Z] [INFO] Syncing scenes...
[2025-12-25T13:29:03.913Z] [INFO] Scenes synced: 320 in 11.6s
[2025-12-25T13:29:04.194Z] [INFO] Syncing images...
[2025-12-25T13:29:04.202Z] [INFO] Images synced: 0 in 0.0s
[2025-12-25T13:29:04.330Z] [INFO] Rebuilding user stats after sync...
[2025-12-25T13:29:04.330Z] [INFO] Rebuilding stats for all users
[2025-12-25T13:29:04.356Z] [INFO] Rebuilding stats for user {"userId":4}
[2025-12-25T13:29:04.368Z] [INFO] Stats rebuild complete {"userId":4,"performerCount":0,"studioCount":0,"tagCount":0}
[2025-12-25T13:29:04.368Z] [INFO] Rebuilding stats for user {"userId":1}
[2025-12-25T13:29:33.776Z] [INFO] Stats rebuild complete {"userId":1,"performerCount":274,"studioCount":140,"tagCount":694}
[2025-12-25T13:29:33.777Z] [INFO] Rebuilding stats for user {"userId":2}
[2025-12-25T13:29:37.783Z] [INFO] Stats rebuild complete {"userId":2,"performerCount":407,"studioCount":177,"tagCount":886}
[2025-12-25T13:29:37.783Z] [INFO] Rebuilding stats for user {"userId":3}
[2025-12-25T13:29:39.310Z] [INFO] Stats rebuild complete {"userId":3,"performerCount":11,"studioCount":7,"tagCount":121}
[2025-12-25T13:29:39.310Z] [INFO] All stats rebuild complete {"userCount":4}
[2025-12-25T13:29:39.310Z] [INFO] User stats rebuild complete
[2025-12-25T13:29:39.310Z] [INFO] Incremental sync completed {"durationMs":1060980,"results":[{"type":"tag","synced":10},{"type":"studio","synced":18},{"type":"performer","synced":1947},{"type":"group","synced":5},{"type":"gallery","synced":0},{"type":"scene","synced":320},{"type":"image","synced":0}]}
[2025-12-25T14:11:57.377Z] [INFO] Scheduled incremental sync triggered
[2025-12-25T14:11:57.740Z] [INFO] Starting incremental sync {"since":"2025-12-17T08:18:12.430Z"}
[2025-12-25T14:11:57.741Z] [INFO] Syncing tags...
[2025-12-25T14:12:00.768Z] [INFO] Tags synced: 10 in 3.0s
[2025-12-25T14:12:00.984Z] [INFO] Syncing studios...
[2025-12-25T14:12:03.075Z] [INFO] Studios synced: 18 in 2.1s
[2025-12-25T14:12:03.359Z] [INFO] Syncing performers...
[2025-12-25T14:23:15.618Z] [INFO] Performers synced: 1,947 in 672.3s
[2025-12-25T14:23:15.870Z] [INFO] Syncing groups...
[2025-12-25T14:23:17.859Z] [INFO] Groups synced: 5 in 2.0s
[2025-12-25T14:23:18.000Z] [INFO] Syncing galleries...
[2025-12-25T14:23:18.006Z] [INFO] Galleries synced: 0 in 0.0s
[2025-12-25T14:23:18.217Z] [INFO] Syncing scenes...
[2025-12-25T14:23:30.658Z] [INFO] Scenes synced: 320 in 12.4s
[2025-12-25T14:23:30.992Z] [INFO] Syncing images...
[2025-12-25T14:23:30.999Z] [INFO] Images synced: 0 in 0.0s
[2025-12-25T14:23:31.204Z] [INFO] Rebuilding user stats after sync...
[2025-12-25T14:23:31.204Z] [INFO] Rebuilding stats for all users
[2025-12-25T14:23:31.240Z] [INFO] Rebuilding stats for user {"userId":4}
[2025-12-25T14:23:31.266Z] [INFO] Stats rebuild complete {"userId":4,"performerCount":0,"studioCount":0,"tagCount":0}
[2025-12-25T14:23:31.266Z] [INFO] Rebuilding stats for user {"userId":1}
[2025-12-25T14:24:04.324Z] [INFO] Stats rebuild complete {"userId":1,"performerCount":274,"studioCount":140,"tagCount":694}
[2025-12-25T14:24:04.324Z] [INFO] Rebuilding stats for user {"userId":2}
[2025-12-25T14:24:08.658Z] [INFO] Stats rebuild complete {"userId":2,"performerCount":407,"studioCount":177,"tagCount":886}
[2025-12-25T14:24:08.658Z] [INFO] Rebuilding stats for user {"userId":3}
[2025-12-25T14:24:10.070Z] [INFO] Stats rebuild complete {"userId":3,"performerCount":11,"studioCount":7,"tagCount":121}
[2025-12-25T14:24:10.070Z] [INFO] All stats rebuild complete {"userCount":4}
[2025-12-25T14:24:10.070Z] [INFO] User stats rebuild complete
[2025-12-25T14:24:10.070Z] [INFO] Incremental sync completed {"durationMs":732682,"results":[{"type":"tag","synced":10},{"type":"studio","synced":18},{"type":"performer","synced":1947},{"type":"group","synced":5},{"type":"gallery","synced":0},{"type":"scene","synced":320},{"type":"image","synced":0}]}
[2025-12-25T15:11:57.386Z] [INFO] Scheduled incremental sync triggered
[2025-12-25T15:11:57.825Z] [INFO] Starting incremental sync {"since":"2025-12-17T08:18:12.430Z"}
[2025-12-25T15:11:57.825Z] [INFO] Syncing tags...
[2025-12-25T15:12:00.206Z] [INFO] Tags synced: 10 in 2.4s
[2025-12-25T15:12:00.475Z] [INFO] Syncing studios...
[2025-12-25T15:12:02.384Z] [INFO] Studios synced: 18 in 1.9s
[2025-12-25T15:12:02.544Z] [INFO] Syncing performers...
[2025-12-25T15:23:05.825Z] [INFO] Performers synced: 1,947 in 663.3s
[2025-12-25T15:23:05.998Z] [INFO] Syncing groups...
[2025-12-25T15:23:07.217Z] [INFO] Groups synced: 5 in 1.2s
[2025-12-25T15:23:07.351Z] [INFO] Syncing galleries...
[2025-12-25T15:23:07.357Z] [INFO] Galleries synced: 0 in 0.0s
[2025-12-25T15:23:07.498Z] [INFO] Syncing scenes...
[2025-12-25T15:23:19.417Z] [INFO] Scenes synced: 320 in 11.9s
[2025-12-25T15:23:19.611Z] [INFO] Syncing images...
[2025-12-25T15:23:19.619Z] [INFO] Images synced: 0 in 0.0s
[2025-12-25T15:23:19.777Z] [INFO] Rebuilding user stats after sync...
[2025-12-25T15:23:19.777Z] [INFO] Rebuilding stats for all users
[2025-12-25T15:23:19.803Z] [INFO] Rebuilding stats for user {"userId":4}
[2025-12-25T15:23:19.813Z] [INFO] Stats rebuild complete {"userId":4,"performerCount":0,"studioCount":0,"tagCount":0}
[2025-12-25T15:23:19.814Z] [INFO] Rebuilding stats for user {"userId":1}
[2025-12-25T15:23:51.587Z] [INFO] Stats rebuild complete {"userId":1,"performerCount":274,"studioCount":140,"tagCount":694}
[2025-12-25T15:23:51.587Z] [INFO] Rebuilding stats for user {"userId":2}
[2025-12-25T15:23:55.627Z] [INFO] Stats rebuild complete {"userId":2,"performerCount":407,"studioCount":177,"tagCount":886}
[2025-12-25T15:23:55.627Z] [INFO] Rebuilding stats for user {"userId":3}
[2025-12-25T15:23:57.003Z] [INFO] Stats rebuild complete {"userId":3,"performerCount":11,"studioCount":7,"tagCount":121}
[2025-12-25T15:23:57.003Z] [INFO] All stats rebuild complete {"userCount":4}
[2025-12-25T15:23:57.004Z] [INFO] User stats rebuild complete
[2025-12-25T15:23:57.004Z] [INFO] Incremental sync completed {"durationMs":719612,"results":[{"type":"tag","synced":10},{"type":"studio","synced":18},{"type":"performer","synced":1947},{"type":"group","synced":5},{"type":"gallery","synced":0},{"type":"scene","synced":320},{"type":"image","synced":0}]}
```
See this line
```
[2025-12-25T15:23:05.825Z] [INFO] Performers synced: 1,947 in 663.3s
```
is happening everytime which causes peek to be unavailable for 10 mins.

2. I have excluded some tags via "Content restriction".
In Content Restriction, i have set "Tags" to "EXCLUDE" and selected a tag.
This was working fine as far as i remember in previous version. But now it's not excluding all scenes.
Looking at the logs, it seems, it's truncating exclusions to only 500 items.
```
[2025-12-25T16:35:55.543Z] [INFO] findScenes: using SQL query builder path
[2025-12-25T16:35:55.552Z] [INFO] getExcludedSceneIds: computed 1531 exclusions in 9ms
[2025-12-25T16:35:55.555Z] [WARN] Exclusion set truncated to 500 items {"originalSize":1531}
[2025-12-25T16:35:55.555Z] [INFO] SceneQueryBuilder.execute {"whereClauseCount":2,"excludedCount":1531,"sort":"random","sortDirection":"DESC","paramCount":504}
[2025-12-25T16:35:55.663Z] [INFO] SceneQueryBuilder.execute complete {"queryTimeMs":111,"breakdown":{"queryMs":56,"countMs":25,"transformMs":1,"relationsMs":25},"resultCount":24,"total":4500}
[2025-12-25T16:35:55.663Z] [INFO] findScenes complete (SQL path) {"totalTimeMs":120,"resultCount":24,"total":4500}
```

3. I have set default sort for scenes as "Random". So ideally everytime i open scenes tab, it should show random scenes.
But it's not. It's showing exactly the same items in a specific order everytime.
```
[2025-12-25T16:39:43.857Z] [INFO] findScenes: using SQL query builder path
[2025-12-25T16:39:43.867Z] [INFO] getExcludedSceneIds: computed 1531 exclusions in 10ms
[2025-12-25T16:39:43.867Z] [WARN] Exclusion set truncated to 500 items {"originalSize":1531}
[2025-12-25T16:39:43.868Z] [INFO] SceneQueryBuilder.execute {"whereClauseCount":2,"excludedCount":1531,"sort":"random","sortDirection":"DESC","paramCount":504}
[2025-12-25T16:39:43.951Z] [INFO] SceneQueryBuilder.execute complete {"queryTimeMs":84,"breakdown":{"queryMs":37,"countMs":16,"transformMs":1,"relationsMs":29},"resultCount":24,"total":4500}
[2025-12-25T16:39:43.951Z] [INFO] findScenes complete (SQL path) {"totalTimeMs":94,"resultCount":24,"total":4500}
```
in the logs above, sort is "random", but it's showing the exact same order everytime.

Hi, after updating to v3.0.2,
1. My performer sync is taking almost twice as long now.
```
[2025-12-29T07:02:07.673Z] [INFO] Performers synced: 1,959 in 1388.9s
```
I tried various full sync, and it's takes 1300-1400s instead of previous around 600s even though it's the same number of performers. And for that 20 min performer sync, both my peek and stash docker containers have almost negligible cpu/memory usage. So something is causing huge delay in performer sync i think.

2. Also, all my full sync are failing to sync scenes now
```
[2025-12-29T07:03:00.820Z] [INFO] Syncing scenes...
[2025-12-29T07:03:09.617Z] [ERROR] Full sync failed {"error":"\nInvalid `prisma.scenePerformer.deleteMany()` invocation:\n\n\nSocket timeout (the database failed to respond to a query within the configured timeout)."}
[2025-12-29T07:03:09.617Z] [ERROR] Startup full sync failed {"error":"\nInvalid `prisma.scenePerformer.deleteMany()` invocation:\n\n\nSocket timeout (the database failed to respond to a query within the configured timeout)."}
```

I restarted both peek and stash and the issue remains.
Sync gets completed in incremental sync, but doing a manual full sync still fails with the above error.

3. The timestamps for incremental sync are still not updating,

```
[2025-12-30T15:33:03.167Z] [INFO] Scheduled incremental sync triggered
[2025-12-30T15:33:03.176Z] [INFO] Starting incremental sync with per-entity timestamps...
[2025-12-30T15:33:03.590Z] [INFO] tag: syncing changes since 2025-12-29T09:22:16+05:30
[2025-12-30T15:33:03.590Z] [INFO] Syncing tags...
[2025-12-30T15:33:04.663Z] [INFO] Tags synced: 1 in 1.1s
[2025-12-30T15:33:04.893Z] [INFO] studio: syncing changes since 2025-12-27T14:32:52+05:30
[2025-12-30T15:33:04.893Z] [INFO] Syncing studios...
[2025-12-30T15:33:05.299Z] [INFO] Studios synced: 1 in 0.4s
[2025-12-30T15:33:05.505Z] [INFO] performer: syncing changes since 2025-12-29T17:45:12+05:30
[2025-12-30T15:33:05.506Z] [INFO] Syncing performers...
[2025-12-30T15:33:05.849Z] [INFO] Performers synced: 1 in 0.3s
[2025-12-30T15:33:06.049Z] [INFO] group: syncing changes since 2025-12-24T16:49:41+05:30
[2025-12-30T15:33:06.049Z] [INFO] Syncing groups...
[2025-12-30T15:33:06.840Z] [INFO] Groups synced: 1 in 0.8s
[2025-12-30T15:33:07.059Z] [INFO] gallery: No previous sync, syncing all
[2025-12-30T15:33:07.059Z] [INFO] Syncing galleries...
[2025-12-30T15:33:07.064Z] [INFO] Galleries synced: 0 in 0.0s
[2025-12-30T15:33:07.342Z] [INFO] scene: syncing changes since 2025-12-30T16:22:34+05:30
[2025-12-30T15:33:07.342Z] [INFO] Syncing scenes...
[2025-12-30T15:33:09.866Z] [INFO] Scenes synced: 1 in 2.5s
[2025-12-30T15:33:10.162Z] [INFO] image: No previous sync, syncing all
[2025-12-30T15:33:10.162Z] [INFO] Syncing images...
[2025-12-30T15:33:10.167Z] [INFO] Images synced: 0 in 0.0s
[2025-12-30T15:33:10.169Z] [INFO] Rebuilding user stats after sync...
[2025-12-30T15:33:10.169Z] [INFO] Rebuilding stats for all users
[2025-12-30T15:33:10.171Z] [INFO] Rebuilding stats for user {"userId":4}
[2025-12-30T15:33:10.234Z] [INFO] Stats rebuild complete {"userId":4,"performerCount":0,"studioCount":0,"tagCount":0}
[2025-12-30T15:33:10.234Z] [INFO] Rebuilding stats for user {"userId":1}
[2025-12-30T15:33:38.254Z] [INFO] Stats rebuild complete {"userId":1,"performerCount":274,"studioCount":140,"tagCount":694}
[2025-12-30T15:33:38.254Z] [INFO] Rebuilding stats for user {"userId":2}
[2025-12-30T15:33:42.372Z] [INFO] Stats rebuild complete {"userId":2,"performerCount":407,"studioCount":177,"tagCount":886}
[2025-12-30T15:33:42.372Z] [INFO] Rebuilding stats for user {"userId":3}
[2025-12-30T15:33:43.793Z] [INFO] Stats rebuild complete {"userId":3,"performerCount":11,"studioCount":7,"tagCount":121}
[2025-12-30T15:33:43.793Z] [INFO] All stats rebuild complete {"userCount":4}
[2025-12-30T15:33:43.793Z] [INFO] User stats rebuild complete
[2025-12-30T15:33:43.793Z] [INFO] Incremental sync completed {"durationMs":40617,"results":[{"type":"tag","synced":1},{"type":"studio","synced":1},{"type":"performer","synced":1},{"type":"group","synced":1},{"type":"gallery","synced":0},{"type":"scene","synced":1},{"type":"image","synced":0}]}
[2025-12-30T16:33:03.183Z] [INFO] Scheduled incremental sync triggered
[2025-12-30T16:33:03.189Z] [INFO] Starting incremental sync with per-entity timestamps...
[2025-12-30T16:33:03.639Z] [INFO] tag: syncing changes since 2025-12-29T09:22:16+05:30
[2025-12-30T16:33:03.640Z] [INFO] Syncing tags...
[2025-12-30T16:33:04.116Z] [INFO] Tags synced: 1 in 0.5s
[2025-12-30T16:33:04.424Z] [INFO] studio: syncing changes since 2025-12-27T14:32:52+05:30
[2025-12-30T16:33:04.424Z] [INFO] Syncing studios...
[2025-12-30T16:33:04.720Z] [INFO] Studios synced: 1 in 0.3s
[2025-12-30T16:33:04.933Z] [INFO] performer: syncing changes since 2025-12-29T17:45:12+05:30
[2025-12-30T16:33:04.933Z] [INFO] Syncing performers...
[2025-12-30T16:33:05.386Z] [INFO] Performers synced: 1 in 0.5s
[2025-12-30T16:33:05.588Z] [INFO] group: syncing changes since 2025-12-24T16:49:41+05:30
[2025-12-30T16:33:05.588Z] [INFO] Syncing groups...
[2025-12-30T16:33:06.401Z] [INFO] Groups synced: 1 in 0.8s
[2025-12-30T16:33:06.619Z] [INFO] gallery: No previous sync, syncing all
[2025-12-30T16:33:06.620Z] [INFO] Syncing galleries...
[2025-12-30T16:33:06.623Z] [INFO] Galleries synced: 0 in 0.0s
[2025-12-30T16:33:06.820Z] [INFO] scene: syncing changes since 2025-12-30T16:22:34+05:30
[2025-12-30T16:33:06.821Z] [INFO] Syncing scenes...
[2025-12-30T16:33:10.481Z] [INFO] Scenes synced: 2 in 3.7s
[2025-12-30T16:33:10.613Z] [INFO] image: No previous sync, syncing all
[2025-12-30T16:33:10.613Z] [INFO] Syncing images...
[2025-12-30T16:33:10.619Z] [INFO] Images synced: 0 in 0.0s
[2025-12-30T16:33:10.756Z] [INFO] Rebuilding user stats after sync...
[2025-12-30T16:33:10.757Z] [INFO] Rebuilding stats for all users
[2025-12-30T16:33:10.760Z] [INFO] Rebuilding stats for user {"userId":4}
[2025-12-30T16:33:10.839Z] [INFO] Stats rebuild complete {"userId":4,"performerCount":0,"studioCount":0,"tagCount":0}
[2025-12-30T16:33:10.839Z] [INFO] Rebuilding stats for user {"userId":1}
[2025-12-30T16:33:39.176Z] [INFO] Stats rebuild complete {"userId":1,"performerCount":274,"studioCount":140,"tagCount":694}
[2025-12-30T16:33:39.176Z] [INFO] Rebuilding stats for user {"userId":2}
[2025-12-30T16:33:43.079Z] [INFO] Stats rebuild complete {"userId":2,"performerCount":407,"studioCount":177,"tagCount":886}
[2025-12-30T16:33:43.080Z] [INFO] Rebuilding stats for user {"userId":3}
[2025-12-30T16:33:44.429Z] [INFO] Stats rebuild complete {"userId":3,"performerCount":11,"studioCount":7,"tagCount":121}
[2025-12-30T16:33:44.429Z] [INFO] All stats rebuild complete {"userCount":4}
[2025-12-30T16:33:44.429Z] [INFO] User stats rebuild complete
[2025-12-30T16:33:44.429Z] [INFO] Incremental sync completed {"durationMs":41240,"results":[{"type":"tag","synced":1},{"type":"studio","synced":1},{"type":"performer","synced":1},{"type":"group","synced":1},{"type":"gallery","synced":0},{"type":"scene","synced":2},{"type":"image","synced":0}]}
[2025-12-30T17:33:03.209Z] [INFO] Scheduled incremental sync triggered
[2025-12-30T17:33:03.244Z] [INFO] Starting incremental sync with per-entity timestamps...
[2025-12-30T17:33:03.383Z] [INFO] tag: syncing changes since 2025-12-29T09:22:16+05:30
[2025-12-30T17:33:03.384Z] [INFO] Syncing tags...
[2025-12-30T17:33:04.484Z] [INFO] Tags synced: 1 in 1.1s
[2025-12-30T17:33:04.699Z] [INFO] studio: syncing changes since 2025-12-27T14:32:52+05:30
[2025-12-30T17:33:04.699Z] [INFO] Syncing studios...
[2025-12-30T17:33:04.985Z] [INFO] Studios synced: 1 in 0.3s
[2025-12-30T17:33:05.329Z] [INFO] performer: syncing changes since 2025-12-29T17:45:12+05:30
[2025-12-30T17:33:05.329Z] [INFO] Syncing performers...
[2025-12-30T17:33:05.687Z] [INFO] Performers synced: 1 in 0.4s
[2025-12-30T17:33:05.896Z] [INFO] group: syncing changes since 2025-12-24T16:49:41+05:30
[2025-12-30T17:33:05.897Z] [INFO] Syncing groups...
[2025-12-30T17:33:06.744Z] [INFO] Groups synced: 1 in 0.8s
[2025-12-30T17:33:06.964Z] [INFO] gallery: No previous sync, syncing all
[2025-12-30T17:33:06.965Z] [INFO] Syncing galleries...
[2025-12-30T17:33:06.975Z] [INFO] Galleries synced: 0 in 0.0s
[2025-12-30T17:33:07.351Z] [INFO] scene: syncing changes since 2025-12-30T21:35:35+05:30
[2025-12-30T17:33:07.351Z] [INFO] Syncing scenes...
[2025-12-30T17:33:10.110Z] [INFO] Scenes synced: 1 in 2.8s
[2025-12-30T17:33:10.337Z] [INFO] image: No previous sync, syncing all
[2025-12-30T17:33:10.337Z] [INFO] Syncing images...
[2025-12-30T17:33:10.342Z] [INFO] Images synced: 0 in 0.0s
[2025-12-30T17:33:10.477Z] [INFO] Rebuilding user stats after sync...
[2025-12-30T17:33:10.477Z] [INFO] Rebuilding stats for all users
[2025-12-30T17:33:10.480Z] [INFO] Rebuilding stats for user {"userId":4}
[2025-12-30T17:33:10.695Z] [INFO] Stats rebuild complete {"userId":4,"performerCount":0,"studioCount":0,"tagCount":0}
[2025-12-30T17:33:10.695Z] [INFO] Rebuilding stats for user {"userId":1}
[2025-12-30T17:33:38.645Z] [INFO] Stats rebuild complete {"userId":1,"performerCount":274,"studioCount":140,"tagCount":694}
[2025-12-30T17:33:38.645Z] [INFO] Rebuilding stats for user {"userId":2}
[2025-12-30T17:33:42.796Z] [INFO] Stats rebuild complete {"userId":2,"performerCount":407,"studioCount":177,"tagCount":886}
[2025-12-30T17:33:42.796Z] [INFO] Rebuilding stats for user {"userId":3}
[2025-12-30T17:33:44.073Z] [INFO] Stats rebuild complete {"userId":3,"performerCount":11,"studioCount":7,"tagCount":121}
[2025-12-30T17:33:44.073Z] [INFO] All stats rebuild complete {"userCount":4}
[2025-12-30T17:33:44.073Z] [INFO] User stats rebuild complete
[2025-12-30T17:33:44.073Z] [INFO] Incremental sync completed {"durationMs":40829,"results":[{"type":"tag","synced":1},{"type":"studio","synced":1},{"type":"performer","synced":1},{"type":"group","synced":1},{"type":"gallery","synced":0},{"type":"scene","synced":1},{"type":"image","synced":0}]}
[2025-12-30T18:33:03.215Z] [INFO] Scheduled incremental sync triggered
[2025-12-30T18:33:03.227Z] [INFO] Starting incremental sync with per-entity timestamps...
[2025-12-30T18:33:03.401Z] [INFO] tag: syncing changes since 2025-12-29T09:22:16+05:30
[2025-12-30T18:33:03.401Z] [INFO] Syncing tags...
[2025-12-30T18:33:03.874Z] [INFO] Tags synced: 1 in 0.5s
[2025-12-30T18:33:04.422Z] [INFO] studio: syncing changes since 2025-12-27T14:32:52+05:30
[2025-12-30T18:33:04.422Z] [INFO] Syncing studios...
[2025-12-30T18:33:04.702Z] [INFO] Studios synced: 1 in 0.3s
[2025-12-30T18:33:04.893Z] [INFO] performer: syncing changes since 2025-12-29T17:45:12+05:30
[2025-12-30T18:33:04.893Z] [INFO] Syncing performers...
[2025-12-30T18:33:05.343Z] [INFO] Performers synced: 1 in 0.5s
[2025-12-30T18:33:05.548Z] [INFO] group: syncing changes since 2025-12-24T16:49:41+05:30
[2025-12-30T18:33:05.548Z] [INFO] Syncing groups...
[2025-12-30T18:33:06.342Z] [INFO] Groups synced: 1 in 0.8s
[2025-12-30T18:33:06.524Z] [INFO] gallery: No previous sync, syncing all
[2025-12-30T18:33:06.524Z] [INFO] Syncing galleries...
[2025-12-30T18:33:06.532Z] [INFO] Galleries synced: 0 in 0.0s
[2025-12-30T18:33:06.724Z] [INFO] scene: syncing changes since 2025-12-30T21:35:35+05:30
[2025-12-30T18:33:06.724Z] [INFO] Syncing scenes...
[2025-12-30T18:33:10.206Z] [INFO] Scenes synced: 2 in 3.5s
[2025-12-30T18:33:10.340Z] [INFO] image: No previous sync, syncing all
[2025-12-30T18:33:10.340Z] [INFO] Syncing images...
[2025-12-30T18:33:10.345Z] [INFO] Images synced: 0 in 0.0s
[2025-12-30T18:33:10.348Z] [INFO] Rebuilding user stats after sync...
[2025-12-30T18:33:10.348Z] [INFO] Rebuilding stats for all users
[2025-12-30T18:33:10.350Z] [INFO] Rebuilding stats for user {"userId":4}
[2025-12-30T18:33:10.408Z] [INFO] Stats rebuild complete {"userId":4,"performerCount":0,"studioCount":0,"tagCount":0}
[2025-12-30T18:33:10.408Z] [INFO] Rebuilding stats for user {"userId":1}
[2025-12-30T18:33:43.313Z] [INFO] Stats rebuild complete {"userId":1,"performerCount":274,"studioCount":140,"tagCount":694}
[2025-12-30T18:33:43.313Z] [INFO] Rebuilding stats for user {"userId":2}
[2025-12-30T18:33:47.309Z] [INFO] Stats rebuild complete {"userId":2,"performerCount":407,"studioCount":177,"tagCount":886}
[2025-12-30T18:33:47.309Z] [INFO] Rebuilding stats for user {"userId":3}
[2025-12-30T18:33:48.591Z] [INFO] Stats rebuild complete {"userId":3,"performerCount":11,"studioCount":7,"tagCount":121}
[2025-12-30T18:33:48.591Z] [INFO] All stats rebuild complete {"userCount":4}
[2025-12-30T18:33:48.591Z] [INFO] User stats rebuild complete
[2025-12-30T18:33:48.591Z] [INFO] Incremental sync completed {"durationMs":45364,"results":[{"type":"tag","synced":1},{"type":"studio","synced":1},{"type":"performer","synced":1},{"type":"group","synced":1},{"type":"gallery","synced":0},{"type":"scene","synced":2},{"type":"image","synced":0}]}
```
In the above logs, incremental sync took place 4 times, and it's still stuck at previous timestamp.

--- a/server/services/StashSyncService.ts
+++ b/server/services/StashSyncService.ts
@@ -48,12 +48,26 @@ const BATCH_SIZE = 500; // Number of entities to fetch per page
  *
  * Stash expects timestamps without timezone suffix. It interprets all timestamps as local time.
  * We store raw timestamp strings from Stash and strip the timezone when querying.
+ *
+ * Additionally, we add .999 milliseconds to handle Stash's sub-second precision.
+ * Stash stores timestamps with sub-second precision internally but returns them truncated
+ * to seconds in API responses. Without this adjustment, querying `> 19:41:58` would still
+ * match an entity with actual timestamp `19:41:58.500`, causing infinite re-syncs.
+ * Adding .999 ensures we skip all entities within that second.
  */
 function formatTimestampForStash(timestamp: string): string {
   // Strip the timezone suffix to get the local time portion
   // "2025-12-28T09:47:03-08:00" -> "2025-12-28T09:47:03"
   // "2025-12-28T09:47:03Z" -> "2025-12-28T09:47:03"
-  return timestamp.replace(/([+-]\d{2}:\d{2}|Z)$/, "");
+  const withoutTz = timestamp.replace(/([+-]\d{2}:\d{2}|Z)$/, "");
+
+  // Add .999 milliseconds to handle sub-second precision
+  // "2025-12-28T09:47:03" -> "2025-12-28T09:47:03.999"
+  // If already has milliseconds, replace them with .999
+  if (/\.\d+$/.test(withoutTz)) {
+    return withoutTz.replace(/\.\d+$/, ".999");
+  }
+  return `${withoutTz}.999`;
 }
 
 /**

--- a/server/services/__tests__/StashSyncService.unit.test.ts
+++ b/server/services/__tests__/StashSyncService.unit.test.ts
@@ -5,6 +5,35 @@
  */
 import { describe, it, expect, vi, beforeEach } from "vitest";
 
+// Test the formatTimestampForStash logic directly (re-implemented here for testing)
+// This mirrors the function in StashSyncService.ts
+function formatTimestampForStash(timestamp: string): string {
+  const withoutTz = timestamp.replace(/([+-]\d{2}:\d{2}|Z)$/, "");
+  if (/\.\d+$/.test(withoutTz)) {
+    return withoutTz.replace(/\.\d+$/, ".999");
+  }
+  return `${withoutTz}.999`;
+}
+
+describe("formatTimestampForStash", () => {
+  it("should strip timezone and add .999 milliseconds", () => {
+    expect(formatTimestampForStash("2025-12-18T19:41:58-08:00")).toBe("2025-12-18T19:41:58.999");
+    expect(formatTimestampForStash("2025-12-28T09:47:03+05:30")).toBe("2025-12-28T09:47:03.999");
+    expect(formatTimestampForStash("2025-12-28T09:47:03Z")).toBe("2025-12-28T09:47:03.999");
+  });
+
+  it("should replace existing milliseconds with .999", () => {
+    expect(formatTimestampForStash("2025-12-18T19:41:58.123-08:00")).toBe("2025-12-18T19:41:58.999");
+    expect(formatTimestampForStash("2025-12-18T19:41:58.5-08:00")).toBe("2025-12-18T19:41:58.999");
+    expect(formatTimestampForStash("2025-12-18T19:41:58.000Z")).toBe("2025-12-18T19:41:58.999");
+  });
+
+  it("should handle timestamps without timezone", () => {
+    expect(formatTimestampForStash("2025-12-18T19:41:58")).toBe("2025-12-18T19:41:58.999");
+    expect(formatTimestampForStash("2025-12-18T19:41:58.500")).toBe("2025-12-18T19:41:58.999");
+  });
+});
+
 // Mock prisma before importing the service
 const mockPrisma = {
   syncState: {

